### PR TITLE
Adding support for multi-bin records

### PR DESCRIPTION
**Pull request by drew-patel**
Hello folks are AppsFlyer,

Thanks for creating this library. It introduced me to Aerospike and got me to play around with it.

I was considering using it for a project and wanted to create secondary indices which require data to be stored as `bins` within `records`. 

This PR introduces `bin` creation via Clojure maps. Each key-value pair in the map is converted to a `bin` and stored in a `record`. There were some issues with `true`, `false` and `nil` values being stored at the top level for a `bin`. I have provided a solution for storing these values in the database and getting them back as they were. Aerospike's documentation suggests setting `bin` values to _null_ to remove them from the database but I did not want to go down that path. 

I have also added the ability to modify stored records by adding or removing `bins`. The current solution does not have a `generation` argument that your single `bin` update function has. That may be something that you want to provide guidance on.

Please let me what other changes you would like made to this PR. Once it is in a state that you are happy with, I would like to contribute by adding a new `query.clj` namespace that would have the following functionality:

- create/drop secondary index
- querying `records` by `bins`
 
Thanks again!
- Drew

### CODE REVIEW

**Comment by barkanido** [comment]
Thanks a lot for the contribution @drew-patel !

I will review your code. 
Generally, it seems like you have done a pretty good job. We will review and merge it.

A few concerns before we go to the nits:
1. I would prefer not to add extra APIs for the "binned" versions. I would prefer either that the user will explicitly tell the library if she wants a single bin or more (this is how the [Java client works](https://www.aerospike.com/apidocs/java/com/aerospike/client/AerospikeClient.html#put-com.aerospike.client.async.EventLoop-com.aerospike.client.listener.WriteListener-com.aerospike.client.policy.WritePolicy-com.aerospike.client.Key-com.aerospike.client.Bin...-) basically) -or- That the library will perform a simple clear decision with the least-surprising-default (based on the `data` argument).
2. I am worried about the way you chose to remove bins. Since this is not the recommended Aeropsike way, the path you took involves 2 API calls to the DB. Although this is logically true, it has poor performance and it is surprising to users. With the first issue solved, this will be much simpler to deal with.

The library tries to follow the Java library as much as it can.

WDYT?

**Comment by drew-patel** [comment]
Thanks for the feedback, @barkanido!

I'm in agreement with both points. Since this was my first time contributing to this library, I did not want to edit existing code.

I will explore both options from the first point and see what comes out of it. As you said, that will make dealing with the second point simpler as well.

**Comment by drew-patel** [comment]
Notes after addressing changes:

**Point 1**
- Single and multiple `bin` `records` are now created and retrieved via the same functions
- To a user coming to the library for the first time, it is conveyed in the doc-string for `put` that Clojure maps will default to creating multiple `bins` while any other type of data will result in a single `bin` `record`


**Point 2**
- `add-bins` and `delete-bins` are new functions that only result in one API call to the database now
- To make them work, I had to create a new `write-policy` that explicitly uses `RecordExistsAction/UPDATE_ONLY` 
- I saw that there was already an `update-policy` but didn't want to change it as it may serve another purpose
- Please advise if you do want to keep them separate or merge them together

**Other changes**
- I have to changed the arguments for `get-single` which now include `bin-keys`
- This seemed necessary but has caused a breaking change for anyone using the _4-arity_ version of this fn since `bin-keys` now takes the place of `conf`
- I had to edit all _4-arity_ usages of `get-single` and add `[:all]` as the 4th argument (code + tests)
- Please advise if this is a deal breaker in this scenario 

Thank you for taking the time to review

**Comment by barkanido** [comment]
Thanks @drew-patel !
This change now does not add new APIs and is a lot better. It also, as you rightfully point, breaks `get-single`. So I wonder if there is a way to avoid it. A few options come in mind:
1. Add the `bin-keys` argument as the last argument. cons: a little awkward maybe?
2. Have the `bin-keys` argument as a part of the `conf` map. cons: very awkward IMO.
3. Only add an additional API to `get-single` (`get-partial`?) cons: you already did this in the former version 😢.
4. a combination of 1. and 3. change `get-single` in a non-breaking way and implement `get-partial` by calling the new version of `get-single`. This would be backward compatible and maybe clearer to users.

WDYT?

Bottom line, in this phase, unfortunately breaking the API is too painful. Yes, I know, decisions...

I will go ahead and start reviewing the implementation to prevent back-and-forth.

File: src/aerospike_clj/client.clj [hunk 1]
```diff
@@ -7,14 +7,17 @@
...
+           [com.aerospike.client.policy Policy ClientPolicy RecordExistsAction]
+           (clojure.lang IPersistentMap IPersistentVector)))
 
 (def EPOCH
   ^{:doc "The 0 date reference for returned record TTL"}
   (.getEpochSecond (java.time.Instant/parse "2010-01-01T00:00:00Z")))
 
 (def MAX_KEY_LENGTH (dec (bit-shift-left 1 13)))
 
+(def MAX_BIN_NAME_LENGTH (- (bit-shift-left 1 4) 2)) ;; 14
```
Comment (barkanido) [inline src/aerospike_clj/client.clj]: can be hardcoded `14`

Comment (barkanido) [inline src/aerospike_clj/client.clj]: can be hardcoded `14`

File: src/aerospike_clj/client.clj [hunk 2]
```diff
@@ -121,29 +124,64 @@
...
+          (utils/desanitize-bin-value (get bins ""))
+          (reduce-kv (fn [m k v]
+                       (assoc m (keyword k) (utils/desanitize-bin-value v)))
+            {}
+            bins))
+        ^Integer (.generation ^Record record)
+        ^Integer (.expiration ^Record record)))))
+
+(defn- map->multiple-bins [^IPersistentMap m]
+  (let [bins (for [[k v] m]
```
Comment (barkanido) [inline src/aerospike_clj/client.clj]: this might be slow because of the creation of the lazy sequence by `for`. I think actually creating a vector with `into` + a transducer and then callling `into-array` over it should be faster.

File: src/aerospike_clj/client.clj [hunk 3]
```diff
@@ -121,29 +124,64 @@
...
+  (if (map? data)
+    (map->multiple-bins data)
+    (into-array Bin [^Bin (Bin. "" (utils/sanitize-bin-value data))])))
 
 (defn get-single
   "Returns a single record: `(transcoder AerospikeRecord)`. The default transcoder is `identity`.
   Pass a `:policy` in `conf` to use a non-default `ReadPolicy`"
-  ([db index set-name] (get-single db index set-name {}))
-  ([db index set-name conf]
+  ([db index set-name] (get-single db index set-name [:all] {}))
```
Comment (barkanido) [inline src/aerospike_clj/client.clj]: after we agree on what API we are going to have here (see top-level PR discussion) we can continue review here.

File: src/aerospike_clj/client.clj [hunk 4]
```diff
@@ -121,29 +124,64 @@
...
+         start-time (System/nanoTime)
+         bin-names (mapv name bin-keys)] ;; bin-names can only be stored as strings in Aerospike
      (.get ^AerospikeClient client
            ^EventLoop (.next ^NioEventLoops (:el db))
            (reify-record-listener op-future)
            ^Policy (:policy conf)
-           (create-key (:dbns db) set-name index))
+           (create-key (:dbns db) set-name index)
+           ^"[Lcom.aerospike.client.Bin;" (if (= [:all] bin-keys)
+                                            (if (utils/single-bin? bin-names) bin-names nil)
```
Comment (barkanido) [inline src/aerospike_clj/client.clj]: is passing `nil` here will actually invoke another version of `.get`? in `AerospikeClient`?

File: src/aerospike_clj/client.clj [hunk 5]
```diff
@@ -121,29 +124,64 @@
...
+(defn- map->multiple-bins [^IPersistentMap m]
+  (let [bins (for [[k v] m]
+               (create-bin (name k) (utils/sanitize-bin-value v)))]
+    (into-array Bin bins)))
+
+(defn- data->bins
+  "Function to identify whether `data` will be stored as a single or multiple bin record.
+  Only Clojure maps will default to multiple bins. Nested data structures are supported."
+  [data]
+  (if (map? data)
```
Comment (barkanido) [inline src/aerospike_clj/client.clj:172]: I like this choice of choosing the data type.

File: src/aerospike_clj/client.clj [hunk 6]
```diff
@@ -197,7 +236,10 @@
   "Writes `data` into a record with the key `index`, with the ttl of `expiration` seconds.
   `index` should be string. Pass a function in `(:trascoder conf)` to modify `data` before it
   is sent to the DB.
-  Pass a `WritePolicy` in `(:policy conf)` to uses the non-default policy."
+  Pass a `WritePolicy` in `(:policy conf)` to uses the non-default policy.
```
Comment (barkanido) [inline src/aerospike_clj/client.clj:257]: thanks for expanding the doc!

File: src/aerospike_clj/client.clj [hunk 7]
```diff
@@ -121,29 +124,64 @@
...
+    (into-array Bin [^Bin (Bin. "" (utils/sanitize-bin-value data))])))
 
 (defn get-single
   "Returns a single record: `(transcoder AerospikeRecord)`. The default transcoder is `identity`.
   Pass a `:policy` in `conf` to use a non-default `ReadPolicy`"
-  ([db index set-name] (get-single db index set-name {}))
-  ([db index set-name conf]
+  ([db index set-name] (get-single db index set-name [:all] {}))
+  ([db index set-name ^IPersistentVector bin-keys] (get-single db index set-name bin-keys {}))
+  ([db index set-name ^IPersistentVector bin-keys conf]
```
Comment (barkanido) [inline src/aerospike_clj/client.clj]: Anyway, I feel that although this is clojure we are dealing here with, a bin name should always be a String. This will make the Clojure maps passed in "non-standard" but will reflect well how this is represented by both the Java client and the DB server. It will also save a few conversions back-and-forths from String to Keyword. the `:all` flag can still stay as a keyword even if we change the type of bin names.  WDYT?

File: src/aerospike_clj/client.clj [hunk 8]
```diff
@@ -285,6 +339,30 @@
...
+  (let [client (get-client db index)
+        op-future (d/deferred)
+        start-time (System/nanoTime)
+        bin-names (mapv name bin-keys)]
+    (.put ^AerospikeClient client
+          ^EventLoop (.next ^NioEventLoops (:el db))
+          ^WriteListener (reify-write-listener op-future)
+          ^WritePolicy policy
+          (create-key (:dbns db) set-name index)
+          ^"[Lcom.aerospike.client.Bin;" (into-array Bin (mapv #(set-bin-as-null %) bin-names)))
```
Comment (barkanido) [inline src/aerospike_clj/client.clj]: no need for the extra `#( %)`

File: src/aerospike_clj/client.clj [hunk 9]
```diff
@@ -301,7 +379,7 @@
            op-future (d/deferred)
            start-time (System/nanoTime)]
        (.operate ^AerospikeClient client
-                 ^EventLoop (.next^NioEventLoops (:el db))
+                 ^EventLoop (.next ^NioEventLoops (:el db))
```
Comment (barkanido) [inline src/aerospike_clj/client.clj:399]: 😄

File: src/aerospike_clj/policy.clj [hunk 10]
```diff
@@ -83,6 +83,14 @@
     (set! (.recordExistsAction wp) RecordExistsAction/REPLACE_ONLY)
     wp))
 
+(defn ^WritePolicy update-bins-policy
```
Comment (barkanido) [inline src/aerospike_clj/policy.clj]: I would rename this to `update-only-policy` so that it might be useful in passing to `put` on the rest of the cases. You can keep the doc about the usefulness of this in the multiple bins case.

File: src/aerospike_clj/utils.clj [hunk 11]
```diff
@@ -11,3 +12,34 @@
...
+  [bin-names]
+  (= bin-names [""]))
+
+(defn sanitize-bin-value
+  "Values in nested structures are unaffected and do not need to be sanitized. If,
+  however, `true`, `false` or `nil` exist as the only value in a bin, they need to
+  be sanitized."
+  [bin-value]
+  (->> (conj [] bin-value)
+       (replace (set/map-invert boolean-replacements))
```
Comment (barkanido) [inline src/aerospike_clj/utils.clj]: I think we should compute the invert of this map in compile-time because this can happen for each API call for each bin, which makes it a hot path. You can also create a transducer from this and later use it in other place to optimize iterations.

File: src/aerospike_clj/utils.clj [hunk 12]
```diff
@@ -11,3 +12,34 @@
...
+  multiple bin record."
+  [bin-names]
+  (= bin-names [""]))
+
+(defn sanitize-bin-value
+  "Values in nested structures are unaffected and do not need to be sanitized. If,
+  however, `true`, `false` or `nil` exist as the only value in a bin, they need to
+  be sanitized."
+  [bin-value]
+  (->> (conj [] bin-value)
```
Comment (barkanido) [inline src/aerospike_clj/utils.clj]: why not just `(vecotr bin-value)` ?

File: test/aerospike_clj/client_test.clj [hunk 13]
```diff
@@ -69,14 +69,74 @@
...
+    (is (true? @(client/create *c* K _set data 100)))
+    (testing "clojure maps can be serialized from bins")
+    (let [v @(client/get-single-no-meta *c* K _set)]
+      (is (= (:foo data) (:foo v))) ;; per value it is identical
+      (is (= (:baz data) (:baz v))) ;; true value returns the same after being sanitized/desanitized
+      (is (= (:qux data) (:qux v))) ;; false value returns the same after being sanitized/desanitized
+      (is (= (:quux data) (:quux v))) ;; nil value retuns the same after being sanitized/desanitized
+      (is (= clojure.lang.PersistentArrayMap (type v))) ;; converted back to a Clojure map instead of HashMap
+      (is (true? (map? v)))
+      (testing "using Jackson to recursively create a PersistantHashMap"
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]: this piece of code was part of the tests already because users accidentally put maps in the single-bin versions API, and those were serialized by Java to `HashMap`s. I added it in order to prove that there is a technical way of easily recursively convert it into a Clojure map. So no need for it here.

File: test/aerospike_clj/client_test.clj [hunk 14]
```diff
@@ -69,14 +69,74 @@
...
+          (is (= data parsed))
+          (is (true? (map? parsed))))))))
+
+(deftest get-single-multiple-bins
+  (let [data {:foo [(rand-int 1000)]
+              :bar [(rand-int 1000)]
+              :baz [(rand-int 1000)]}]
+    (is (true? @(client/create *c* K _set data 100)))
+    (testing "bin values can be retrieved individually and all together")
+    (let [v1 @(client/get-single *c* K _set [:foo])
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]: this `let` should be part of the `testing` macro

File: test/aerospike_clj/client_test.clj [hunk 15]
```diff
@@ -69,14 +69,74 @@
...
+
+(deftest adding-bins-to-record
+  (let [data {:foo [(rand-int 1000)]
+              :bar [(rand-int 1000)]
+              :baz [(rand-int 1000)]}
+        new-data {:qux [(rand-int 1000)]}]
+    (is (true? @(client/create *c* K _set data 100)))
+    (is (true? @(client/add-bins *c* K _set new-data 100))) ;; adding value to bin
+    (testing "bin values can be added to existing records")
+    (let [v @(client/get-single-no-meta *c* K _set)]
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]: this `let` should be part of the `testing` macro

File: test/aerospike_clj/client_test.clj [hunk 16]
```diff
@@ -69,14 +69,74 @@
...
+(deftest removing-bins-from-record
+  (let [data {:foo [(rand-int 1000)]
+              :bar [(rand-int 1000)]
+              :baz [(rand-int 1000)]
+              :qux [(rand-int 1000)]}
+        bin-keys [:foo :bar :baz]]
+    (is (true? @(client/create *c* K _set data 100)))
+    (is (true? @(client/delete-bins *c* K _set bin-keys 100))) ;; removing value from bin
+    (testing "bin values can be removed from existing records")
+    (let [v @(client/get-single-no-meta *c* K _set)]
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]: this `let` should be part of the `testing` macro

File: test/aerospike_clj/client_test.clj [hunk 17]
```diff
@@ -69,14 +69,74 @@
...
+(deftest adding-bins-to-record
+  (let [data {:foo [(rand-int 1000)]
+              :bar [(rand-int 1000)]
+              :baz [(rand-int 1000)]}
+        new-data {:qux [(rand-int 1000)]}]
+    (is (true? @(client/create *c* K _set data 100)))
+    (is (true? @(client/add-bins *c* K _set new-data 100))) ;; adding value to bin
+    (testing "bin values can be added to existing records")
+    (let [v @(client/get-single-no-meta *c* K _set)]
+      (is (= v (merge data new-data)))
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]: maybe also test what happens if you call `add-bins` with an existing bin?

File: test/aerospike_clj/client_test.clj [hunk 18]
```diff
@@ -69,14 +69,74 @@
...
+              :bar [(rand-int 1000)]
+              :baz [(rand-int 1000)]
+              :qux [(rand-int 1000)]}
+        bin-keys [:foo :bar :baz]]
+    (is (true? @(client/create *c* K _set data 100)))
+    (is (true? @(client/delete-bins *c* K _set bin-keys 100))) ;; removing value from bin
+    (testing "bin values can be removed from existing records")
+    (let [v @(client/get-single-no-meta *c* K _set)]
+      (is (= v (apply dissoc data bin-keys)))
+      (is (contains? v :qux)))))
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]: you can be more strict here and compare `(keys v)` to `[:qux]`

**Review by barkanido — CHANGES_REQUESTED** [review]

File: src/aerospike_clj/client.clj (hunk 1 above)
Comment (drew-patel) [reply src/aerospike_clj/client.clj]: done

Comment (drew-patel) [reply src/aerospike_clj/client.clj]: same as above

File: src/aerospike_clj/client.clj (hunk 4 above)
Comment (drew-patel) [reply src/aerospike_clj/client.clj]: Yes. It seems that passing `nil` here returns all the `bins` in this record. I'm actually not a fan of the nested `if` statements since it makes this function a bit harder to reason about. 

I'll try to clean it up when I am improving this function but suggestions are welcome

File: src/aerospike_clj/client.clj (hunk 5 above)
Comment (drew-patel) [reply src/aerospike_clj/client.clj:172]: :+1:

File: src/aerospike_clj/client.clj (hunk 6 above)
Comment (drew-patel) [reply src/aerospike_clj/client.clj:257]: :+1:

File: src/aerospike_clj/client.clj (hunk 7 above)
Comment (drew-patel) [reply src/aerospike_clj/client.clj]: I think initially, I did want to do this as well and then chose to default to Clojure keywords. I think there are two options here:
- ensure users only use string keys in this library (error when they don't). Probably better since that is how Aerospike wants it
OR
- convert every key into a string and expect them to call it back as such. Pretty much `str` everything in and out to be safe

File: src/aerospike_clj/client.clj (hunk 8 above)
Comment (drew-patel) [reply src/aerospike_clj/client.clj]: Haha. Thanks!

File: src/aerospike_clj/policy.clj (hunk 10 above)
Comment (drew-patel) [reply src/aerospike_clj/policy.clj]: done

File: src/aerospike_clj/utils.clj (hunk 11 above)
Comment (drew-patel) [reply src/aerospike_clj/utils.clj]: will do

File: src/aerospike_clj/utils.clj (hunk 12 above)
Comment (drew-patel) [reply src/aerospike_clj/utils.clj]: I was getting errors casting `true`, `false` and `nil` to vectors in the REPL. It was either `[bin-value]` or use `conj` instead

File: test/aerospike_clj/client_test.clj (hunk 14 above)
Comment (drew-patel) [reply test/aerospike_clj/client_test.clj]: whoops!

File: test/aerospike_clj/client_test.clj (hunk 17 above)
Comment (drew-patel) [reply test/aerospike_clj/client_test.clj]: :+1:

File: test/aerospike_clj/client_test.clj (hunk 18 above)
Comment (drew-patel) [reply test/aerospike_clj/client_test.clj]: :+1:

File: src/aerospike_clj/utils.clj (hunk 12 above)
Comment (drew-patel) [reply src/aerospike_clj/utils.clj]: Right... will do! I was in an `into` mood and getting errors casting `true`, `false` and `nil` :disappointed:

File: src/aerospike_clj/client.clj (hunk 7 above)
Comment (barkanido) [reply src/aerospike_clj/client.clj]: I'll go for the error option. Less surprising, more perfomant

File: test/aerospike_clj/client_test.clj (hunk 14 above)
Comment (drew-patel) [reply test/aerospike_clj/client_test.clj]: done

File: test/aerospike_clj/client_test.clj (hunk 15 above)
Comment (drew-patel) [reply test/aerospike_clj/client_test.clj]: done

File: test/aerospike_clj/client_test.clj (hunk 16 above)
Comment (drew-patel) [reply test/aerospike_clj/client_test.clj]: done

File: src/aerospike_clj/client.clj (hunk 2 above)
Comment (drew-patel) [reply src/aerospike_clj/client.clj]: done

File: test/aerospike_clj/client_test.clj (hunk 13 above)
Comment (drew-patel) [reply test/aerospike_clj/client_test.clj]: removed

**Comment by drew-patel** [comment]
Thank you for the review, @barkanido 

I have added almost everything that was brought up in the comments. I have closed the smaller conversations and left the larger pieces open for discussion.

- As discussed, `get-single` is back to its original state with the argument for `bins` being added to the end
- Transducers are added. I hope I did an OK job with them
- Only string keys are allowed when storing `bins` in the database. Exception is thrown for everything else
- The nested `if` statements are the only thing I haven't touched at the moment at `client.clj` L182-184 since we may make changes to that function after review

File: src/aerospike_clj/client.clj (hunk 4 above)
Comment (barkanido) [reply src/aerospike_clj/client.clj]: sometimes it is what it is. breaking the inner `if` to separate lines and documenting the `nil case` might help. there is `clojure.core.match` for those things but it is a bit of too much black magic and we care about performance here.

File: src/aerospike_clj/utils.clj [hunk 19]
```diff
@@ -11,3 +12,45 @@
...
+(defn string-keys?
+  "Predicate function to determine whether all keys provided for bins are strings."
+  [bin-names]
+  (every? string? bin-names))
+
+(defn sanitize-bin-value
+  "Values in nested structures are unaffected and do not need to be sanitized. If,
+  however, `true`, `false` or `nil` exist as the only value in a bin, they need to
+  be sanitized."
+  [bin-value]
```
Comment (barkanido) [inline src/aerospike_clj/utils.clj:48]: If `bin-value` is a single (not a sequence) always, why do we need a transducer, to begin with? I think this is my fault for not understanding this earlier. If I understand correctly you use `replace` here  (via transducing) but you never use it on a sequence of bins. So, for a single bin should we could just replace a key with a value from `boolean-replacements` (or the invert of it)?

So, I think `x-bin-convert` is good but `x-sanitize` and `x-desanitize` should not be transducers because they are always 'iterating' over a single bin. WDYT? It might be worth it to [micro benchmark](https://github.com/hugoduncan/criterium) this.

File: test/aerospike_clj/client_test.clj [hunk 20]
```diff
@@ -64,18 +64,73 @@
...
+      (is (= clojure.lang.PersistentArrayMap (type v))))))
+
+(deftest put-multiple-bins-get-clj-map
+  (let [data {"foo" {"bar" [(rand-int 1000)]}
+              "baz" true
+              "qux" false
+              "quuz" nil}]
+    (is (true? @(client/create *c* K _set data 100)))
+    (testing "clojure maps can be serialized from bins")
+    (let [v @(client/get-single-no-meta *c* K _set)]
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]: `let` should be inside `testing` :)

File: test/aerospike_clj/client_test.clj [hunk 21]
```diff
@@ -167,7 +228,7 @@
 (deftest transcoder-failure
   (is (true? @(client/create *c* K _set 1 100)))
   (let [transcoder (fn [_] (throw (Exception. "oh-no")))]
-    (is (thrown-with-msg? Exception #"oh-no" @(client/get-single *c* K _set {:transcoder transcoder})))))
+    (is (thrown-with-msg? Exception #"oh-no" @(client/get-single *c* K _set {:transcoder transcoder} [:all])))))
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]: all additions of `[:all]` to `get-single` are redundant now, right?

**Review by barkanido — COMMENTED** [review]
Thanks, @drew-patel ! We are definitely getting there! A few more nits and I think we are done.

File: test/aerospike_clj/client_test.clj [hunk 22]
```diff
@@ -64,18 +64,73 @@
...
+  (let [data {"foo" [(rand-int 1000)]
+              "bar" [(rand-int 1000)]
+              "baz" [(rand-int 1000)]}
+        new-data {"qux" [(rand-int 1000)]}]
+    (is (true? @(client/create *c* K _set data 100)))
+    (is (true? @(client/add-bins *c* K _set new-data 100))) ;; adding value to bin
+    (testing "bin values can be added to existing records"
+      (let [v @(client/get-single-no-meta *c* K _set)]
+        (is (= v (merge data new-data)))
+        (is (= (contains? v "qux")))))
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]: `java.lang.Exception: = expects more than one argument`

File: test/aerospike_clj/client_test.clj (hunk 20 above)
Comment (drew-patel) [reply test/aerospike_clj/client_test.clj]: Done. I really should be paying more attention to this

File: test/aerospike_clj/client_test.clj (hunk 21 above)
Comment (drew-patel) [reply test/aerospike_clj/client_test.clj]: you are correct! removed them all

File: src/aerospike_clj/client.clj (hunk 4 above)
Comment (drew-patel) [reply src/aerospike_clj/client.clj]: sounds good. added comment and separated inner `if` statement

File: src/aerospike_clj/utils.clj (hunk 19 above)
Comment (drew-patel) [reply src/aerospike_clj/utils.clj:48]: After benchmarks, here are some stats:

> sanitize-bin-value **with transducer**
`Execution time mean : 266.958899 ns`

> desanitize-bin-value **with transducer**
`Execution time mean : 218.256747 ns`

> sanitize-bin-value
`Execution time mean : 358.403094 ns`

> desanitize-bin-value
`Execution time mean : 403.246815 ns`

Seems like transducers do help very slightly _(100-185 ns faster)_. It's negligible but I will leave transducers in for now unless you feel they are overkill and want to remove them.

File: test/aerospike_clj/client_test.clj [hunk 23]
```diff
@@ -64,18 +64,73 @@
...
```
Comment (barkanido) [inline test/aerospike_clj/client_test.clj]:
```suggestion
        (is (true? (contains? v "qux")))))
```

Comment (barkanido) [reply test/aerospike_clj/client_test.clj]: I think that this is the last bit :)

**Comment by barkanido** [comment]
pinging this.

**Comment by drew-patel** [comment]
@barkanido Anything else left to do? I left the last comment for the transducer benchmarks just to see if you wanted them in or not. Let me know if you'd like them removed. 

But yes, you are correct in that we never iterate over a data structure but always a single value. 

Thanks!

**Comment by barkanido** [comment]
@drew-patel. Only fixing the failing test. See suggestion above. I'll then do a final sweep and merge.

**Comment by drew-patel** [comment]
@barkanido Sorry about that. I think I fixed what you were looking for. 

Please let me know if you see anything else after the final sweep or if there was anything else that I missed

**Comment by barkanido** [comment]
1 test is failing with an error:
```
lein test :only aerospike-clj.client-test/adding-bins-to-record

ERROR in (adding-bins-to-record) (client_test.clj:114)
bin values can be added to existing records
expected: (= (contains? v "qux"))
  actual: java.lang.Exception: = expects more than one argument
 at aerospike_clj.client_test$fn__6777$fn__6786.invoke (client_test.clj:114)
```

**Comment by drew-patel** [comment]
@barkanido Thanks for pointing that out. Would you happen to know why I didn't see any test failures on my end or CI? Prior to you showing me that stack trace, all my tests were still passing. Really strange

**Comment by barkanido** [comment]
No idea:). 
Another issue is a disturbing reflection warning that has now popped up and I am a bit afraid it would hurt performance in a noticeable way:
```
Reflection warning, aerospike_clj/client.clj:186:6 - call to method get on com.aerospike.client.AerospikeClient can't be resolved (argument types: com.aerospike.client.async.EventLoop, com.aerospike.client.listener.RecordListener, com.aerospike.client.policy.Policy, com.aerospike.client.Key, unknown).
```
Note the last argument that the compiler insist to not kbow about since albeit the type hint.

**Review by barkanido — APPROVED** [review]
@drew-patel thanks for the hard work. I will go ahead and merge it.
